movici_simulation_core/models/rules/expression.py

```diff
@@ -0,0 +1,347 @@
+    ).setParseAction(lambda: False)
+
+    value = duration | clock_time | number | boolean
+
+    comparison_op = pp.oneOf("== != <= >= < >")
+
+    simtime_var = (
+        pp.Literal("<simtime>") | pp.Literal("<simtime") + pp.Literal(">")
+    ).setParseAction(lambda: ExpressionType.SIMTIME)
+
+    clocktime_var = (
+        pp.Literal("<clocktime>")
+        | pp.Literal("<clocktime") + pp.Literal(">")
+        | pp.CaselessKeyword("clocktime")
+    ).setParseAction(lambda: ExpressionType.CLOCKTIME)
+
+    attribute_name = pp.Combine(
+        pp.Word(pp.alphas + "_", pp.alphanums + "_")
+        + pp.Optional("." + pp.Word(pp.alphas + "_", pp.alphanums + "_"))
+    )
+
+    lhs = simtime_var | clocktime_var | attribute_name
+
+    def make_comparison(tokens: pp.ParseResults) -> Comparison:
+        lhs_val = tokens[0]
+        op_str = tokens[1]
+        rhs_val = tokens[2]
+
+        if isinstance(lhs_val, ExpressionType):
+            expr_type = lhs_val
+            attr_name = None
+        else:
+            expr_type = ExpressionType.ATTRIBUTE
+            attr_name = lhs_val
+
+        op_map = {
+            "==": ComparisonOperator.EQ,
+            "!=": ComparisonOperator.NE,
+            "<": ComparisonOperator.LT,
+            "<=": ComparisonOperator.LE,
+            ">": ComparisonOperator.GT,
+            ">=": ComparisonOperator.GE,
+        }
+
+        return Comparison(
+            expr_type=expr_type,
+            operator=op_map[op_str],
+            value=rhs_val,
+            attribute_name=attr_name,
+        )
+
+    comparison = (lhs + comparison_op + value).setParseAction(make_comparison)
+
+    NOT = pp.CaselessKeyword("NOT") | pp.Literal("!")
+    AND = pp.CaselessKeyword("AND") | pp.Literal("&")
+    OR = pp.CaselessKeyword("OR") | pp.Literal("|")
+
+    bool_expr = pp.infixNotation(
+        comparison,
+        [
+            (NOT, 1, pp.opAssoc.RIGHT, _NotExpr),
+            (AND, 2, pp.opAssoc.LEFT, _BinaryExpr),
+            (OR, 2, pp.opAssoc.LEFT, _BinaryExpr),
+        ],
+    )
+
+    return bool_expr
+
+
+_GRAMMAR = _build_grammar()
+
+
+def parse_condition(expr_string: str) -> ParsedCondition:
+    """Parse a condition expression string.
+
+    Supports simple comparisons and compound boolean expressions:
+
+    - ``"<simtime> == 34h"``
+    - ``"level >= 23"``
+    - ``"level <= 21 & level >= 23"``
+    - ``"(level < 10 | level > 90) & status == true"``
+    - ``"NOT status == true"``
+
+    :param expr_string: Condition expression string
+    :returns: Parsed condition (Comparison or BooleanExpression)
+    :rtype: ParsedCondition
+    :raises ValueError: If the expression cannot be parsed
+    """
+    try:
+        result = _GRAMMAR.parseString(expr_string.strip(), parseAll=True)
+    except pp.ParseException as e:
+        raise ValueError(f"Invalid condition expression: {expr_string!r}") from e
+
+    parsed = result[0]
+
+    if hasattr(parsed, "to_expr"):
+        return parsed.to_expr()
```

Comment:
Be careful with hasattr. It makes the code harder to reason about. Perhaps an instance check works better?

Other option, that I think might be preferable is to not return _NotExpr or _BinaryExpr, but have the to_expr call included in the grammar parsing:

```python
def not_expr(*args, **kwargs):
    return _NotExpr(*args, **kwargs).to_expr()
```
or a similar function and use that as a callback in the grammar